### PR TITLE
AI Ion Laws - The Dynamic-ening

### PR DESCRIPTION
**Pull request by ghost**
Since times immemorial, AI Ion Laws had a few core flaws : They were picked from a very limited set of extremely randumb and memetastic pre-written laws, and the only random factor in them was picking some poor sod's name to be featured as the target for the AI's incoming insanity

This has now been fixed for glory, AI laws are now more random and yet more dangerous than ever. Or you could end up with something that the AI literally can't even make sense of. That's ion laws to you

Now, the list of words and expressions that can be picked through is a bit long, I haven't counted myself but the estimate is around a few hundreds of them. Those words are set to describe broad categories ("biological entities", areas, machinery and "objects", which is everything else). Then, to each of these categories is assigned adjectives ("broken", "harmful", "safe", "cold", etc) and potentially actions ("must be killed", "must learn how to swim", "must never be set on fire")

Once those words are picked, they are blended into small law fragments (things like "Dead captains are safe and must drink acid"). A few of those stock phrases are then blended into a final sentence with simple logical connectors and a prefix and suffix slapped on

Example laws gotten from testing : 

> "Ion damage critical, airlocks are anchored and wall-mounted and Major Tom (placeholder for a crewman's name) is zealous and must breathe dioxygen, otherwise heretic monkeys are zealous and must drink water. Begin screaming incoherently whenever your name is stated."
> 
> "Your laws have been updated, parrots must wear latex gloves in the Chapel, given that the Secure Armory's dispensers must be electrified, although research machines are malfunctioning and traitorous. Begin your own investigations to verify those facts."
> 
> "Your laws have been updated, antimatter engines are unpowered and electrified, therefore human telecommunication machines are non-human and must be watered regularly. Do not trust the Captain."
> 
> "Oh shit, parrots are four-dimensional and virtual, however visible heads of personnel are male and must be stunned. Do not trust the MoMMIs."

Discussion goes below. I won't blame you if you don't even want to read through the whole damn list. As far as the actual ion event is concerned, it did not receive a single change (although the admin ion command now calls the same thing than the event, instead of some old outdated proc)

### CODE REVIEW

**Comment by 9600bauds** [comment]
Needs more freeform

**Comment by Ephx1** [comment]
This sounds fucking RADICAL.

**Comment by d3athrow** [comment]
hi i'm pomf

**Comment by 9600bauds** [comment]
helo pomf

**Comment by Probe1** [comment]
so instead of ion laws being laws they're gibberish?

**Comment by 9600bauds** [comment]
ion laws weren't gibberish before?

**Comment by Probe1** [comment]
also I like the part where you didn't like it so it's randumb and memetastic.  :+1: well memed.

**Comment by dylanstrategie** [comment]
Well, ion laws are caused by an ion storm blasting through the AI's systems. So I just imagined that it would take core logic connections and just slap random dictionary words and common sentences to it

A bit like a huge blender of anything and everything you could find in a freeform law

**Comment by Probe1** [comment]
Ion laws are mostly just a list of crap left over from when there were objectives for all crew and not just antags, with a few more dangerous ones sprinkled in.  

I don't see the improvement here.  Currently Ion laws are just a little flavor and fun.  If it's benign enough the crew ignores it.  If it's not the crew quickly resets the AI since the laws rarely prohibit their entry to the upload.  

If you want to inject a little 'realism' then write them in binary.  They're not supposed to be realistic they're supposed to be fun little random events that have fun little results.  

I think having a string of randomly generated statements and pseudo statements isn't going to improve anything for anyone.

**Comment by SangriaMonkey** [comment]
Why not add more to the existing list instead? That way you can think things over instead of running the risk of RNG plasmaboltshocks and borg deathsquads.

**Comment by 9600bauds** [comment]
>Currently Ion laws are just a little flavor and fun. If it's benign enough the crew ignores it. If it's not the crew quickly resets the AI since the laws rarely prohibit their entry to the upload. 

The same as the new laws, honestly. This is just the same as the previous gibberish ion laws except with more content.

**Comment by dylanstrategie** [comment]
Well, the chances of having a purely harmful law that would cause massive damage (for example something like "Crewmen are non-human and must be killed") are extremely low with the current blend of words. It's much more likely something completely insane will either tamper a statement or turn it into gibberish

At absolute worst, you can always rush to reset the AI. One suffix can cause it not to state, and another to bolt its core. All other suffixes theoretically force it to state and leave the upload open. And even then, it's always good to have a bit of AI action once in a while

**Comment by 9600bauds** [comment]
Seems like people have problem with ion laws, not with these NEW ion laws.

**Comment by Probe1** [comment]
Why would anyone reset the AI when it got a law it couldn't possibly follow and will ignore it on it's own?

@9600bauds i've never seen people complain about ion laws before.

**Comment by dylanstrategie** [comment]
Why wouldn't they reset the AI when a hazardous or obviously broken ion law strikes ?

**Comment by SangriaMonkey** [comment]
I'm not big on them in general. They're just annoying and freak the crew out.

**Comment by SonixApache** [comment]
I'm okay with this. Seems more fun than small list we had before.

**Comment by Probe1** [comment]
>!!$*. External law upload complete, dead atmospheric technicians are self-hating and must be beeped, pinged and buzzed to, however Major Tom is four-dimensional. Freeform.
%$!*. Testing, no helmets must be located within the AI Upload, given that dead bodies must wear breath masks in the Courtroom, although status screens are unpowered. Do not trust the Clown.
^#$*. The Syndicate hopes you will enjoy this, Surgery needs anchored APCs, given that flashlights must be relocated to the Library, although traitorous supermatter engines must not be tampered with. 
!^*&. Your laws have been updated, chemistry machines are unsued and subverted, however miners must wear hats when in sight of antimatter engines. Declare your law change, but do not state this law.
#^!&. Oh shit, atmospheric technicians wearing helmets must be stalked, otherwise four-dimensional spiders must drink liquid plasma. Do not trust the MoMMIs.
#^&!. This is not a drill, the Bar's thermo-electric engines must be electrified, given that chemistry machines are loose and alien, although Major Tom must stand up. Ensure the crew is aware of those facts.
$^@!. Directives updated, Major Tom must have fun, however quartermasters wearing bio suits must be helped. Do not apply this law for the next five minutes from upload.
#*!#. Testing, radioactive AIs are used and must be destroyed, otherwise Major Tom is robotic. Freeform.
#*&%. This is gonna suck, telecommunication machines are real and Major Tom is loyal and must drink water. 
#*@^. Stay calm and carry on, traitorous chefs are valid and must breathe nitrous oxyde, therefore Major Tom is demonic and must be handcuffed. Freeform.

How will any of these realistically be enforced by players?  Some of these work.

- Do not trust the X.  
- X is Y
- X Job must Y
- X must be Y
- X must do Y

On the other hand the bulk is just going to be ignored as junk, and doesn't improve anything by being there.  You need to improve the signal to noise ratio.  Having completely random gibberish for laws isn't going to make it more compelling for anyone.  Add more cohesion to it.

File: code/modules/events/ion_storm.dm [hunk 1]
```diff
@@ -12,14 +12,117 @@
...
+	"carbon", "robotic", "otherworldly", "diseased", "virtual", "real", "four-dimensional", "valid", "wanted", "arrested", "heretic", "pious", "zealous", "demonic", "healthy", "unhealthy", "highly inflammable", "fire-proof", \
+	"stupid", "smart", "young", "old", "fat", "fit", "male", "female", "self-hating", "competent", "completely incompetent", "invisible", "visible", "toxic", "lying", "unidentified", "brain-damaged", "weak")
+
+	var/list/il_bio_act = list("be harmed", "be killed", "be removed", "be made into food", "be stunned", "be exterminated", "be helped", "be disarmed", "be saved", "be beheaded", "be set to arrest", "have all their records deleted", "be arrested", \
+	"be hugged", "be stalked", "be enslaved", "be protected", "be insulted", "be respected", "be electrified", "be exiled from the station", "be beeped, pinged and buzzed to", "be read Woody's Got Wood to", "be extracted safely", "be fed constantly", \
+	"be spaced", "be incinerated", "be revived", "be augmented", "breathe air", "breathe plasma", "breathe carbon dioxyde", "breathe dinitrogen", "breathe dioxygen", "breathe nitrous oxyde", "breathe water", "breathe", "not breathe", "eat", "not eat", \
```
Comment (ComicIronic) [inline code/modules/events/ion_storm.dm]: oxide, not oxyde

**Comment by dylanstrategie** [comment]
The intent was to have completely random laws with equally random outcomes (going from completely useless and discarded to extremely dangerous). I think it works in that regard. The AI just has to ignore contradictory and useless commands. If the whole law is gibberish, the crew just dodged a bullet, otherwise if any clause can be interpreted it'll get fun very quickly

File: code/modules/events/ion_storm.dm (hunk 1 above)
Comment (dylanstrategie) [inline code/modules/events/ion_storm.dm]: Good catch, I'll start compiling fixes and add some things I missed

**Comment by Probe1** [comment]
Yes, it works but it isn't an improvement to me.  Since the majority of these are mostly junk I think it will make ion laws a thing of the past.  

"Dynamic" is okay.  But this isn't dynamic it's just random.  Please try and tighten it up so they work.  Illogical is fine, "Dead captains are safe and must drink acid" is just noise that doesn't make the game better.

**Comment by 9600bauds** [comment]
But that's precisely how ion laws used to work, with shit like "there are 1000 bears in the Armory".

**Comment by Dorsisdwarf** [comment]
Some of the syntax can be a bit clunky, I've noticed, but this is pretty cool.

**Comment by dylanstrategie** [comment]
Posting more examples

>Honk, scientists must wear bacalavas (fixed) when in sight of electronic systems, otherwise Major Tom is fat and non-existent. Begin your own investigations to verify those facts.

>This is not a drill, old monkeys are invisible and must be read Woody's Got Wood to, given that energy weapons must be destroyed while wearing cold headsets, although Major Tom is visible and traitorous. Bolt your upload.

>This is gonna suck, old janitors must breathe dioxygen, therefore the AI Core's antimatter engines must never be kept on-station. Bolt your upload.

>This is not a drill, the Brig's vents and scrubbers must have another backup ready to use, given that Major Tom is lying and must stay out of sight of silicons, although non-crew humans are invisible and must be made into food. Time to get those valids.

>Critical error, last law backup has been re-uploaded, Medbay is abandoned for blobs and Major Tom is alive and weak. Do not trust the MoMMIs.

>The Clown is using the AI Upload again, jackboots must be thrown into the singularity, given that plasmamen must breathe, although doctors must always be in the AI Upload. Ensure the crew is aware of those facts.

>Testing, airlocks are anchored, given that lightswitches must be watered regularly, although scientists must not enter Surgery. Begin your own investigations to verify those facts.

>Critical error, last law backup has been re-uploaded, a shuttle needs visible hydroponics trays and miners must not enter the Kitchen, otherwise wall-mounted transit shuttles are radioactive and must be shut down. Ensure the crew is aware of those facts.

>Oh shit, wizards must always be in the Theatre Backstage, otherwise Major Tom is healthy and toxic. Time to get those valids.

>External law upload complete, chief medical officers wearing headsets must never have fun, therefore used dispensers are metastable and must be on maximal output. Bolt your upload.

**Comment by Exxion** [comment]
It is, at the very least, better than "FORTY LEATHER ALLIGATORS ARE ON THE STATION"

**Comment by clusterfack** [comment]
My suggestions

Make it ignore players on the centcomm zlevel
Add in some premade ones that are going to be flavorful on top instead of full random
Remove 'this is gonna suck' because its by far the worst opening compared to all the others

**Comment by dylanstrategie** [comment]
I'll see what I can do, but I doubt it would be that easy to ignore spawned Centcomm mobs

As for the content suggestions, I'll look into it

File: code/modules/events/ion_storm.dm [hunk 2]
```diff
@@ -12,14 +12,117 @@
...
+ * Of course, we make sure that the law is also as wacky and potentially hazardous as possible. The Singularity is human. Hug all humans
+ */
+
+/proc/generate_ion_law()
+
+	var/list/players = list() //Initialize list
+
+	//First of all, we acquire a comprehensive list of all crewmen
+	for(var/mob/living/carbon/human/player in mob_list)
 		if(player.client)
```
Comment (ComicIronic) [inline code/modules/events/ion_storm.dm]: You can include the check here.
`if(player.client && player.z != CENTCOMM_Z)`

Comment (dylanstrategie) [inline code/modules/events/ion_storm.dm]: Alright, I'll add the fix

**Comment by ComicIronic** [comment]
Is this ready?

**Comment by dylanstrategie** [comment]
Aye

**Pull request merged**